Add articles "Prerelease testing" & "Self-hosting", use new API (#3053)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/how-to/set-preview-channel.md

ms.topic: conceptual
ms.service: microsoft-edge
ms.subservice: webview
ms.date: 02/09/2024
---
# Test upcoming APIs and features
<!-- old title: 
# Switch to a preview channel to test upcoming APIs and features
-->

> <!-- todo: update the article to use the new API -->

Author: update the "Test upcoming APIs and features" article (set-preview-channel.md) to use the new API
@victorhuangwq

Author: article, in Files Changed tab, has had extensive modifs.  assuming shows new api; covers API approach incl WinRT.  
closing.

* [CoreWebView2CreationProperties Class](/dotnet/api/microsoft.web.webview2.winforms.corewebview2creationproperties)

WinRT:
* [CoreWebView2Environment.CreateWithOptionsAsync Method](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environment#createwithoptionsasync)

Win32:
* [CreateCoreWebView2EnvironmentWithOptions](/microsoft-edge/webview2/reference/win32/webview2-idl?#createcorewebview2environmentwithoptions)

##### [Registry key](#tab/registry-key)

1. Open a PowerShell terminal or PowerShell-enabled command prompt.

1. Modify and then run the following command:

   ```reg
   REG ADD <HKLM/HKCU>\Software\Policies\Microsoft\Edge\WebView2\BrowserExecutableFolder /v WebView2APISample.exe /t REG_SZ /d "C:\Users\myname\AppData\Local\Microsoft\Edge SxS\Application\88.0.680.0"
   ```

Replace `WebView2APISample.exe` with your own app executable name or the application user model ID. Replace the sample path with the path to the desired Microsoft Edge preview channel.

##### [Environment variable](#tab/environment-variable)

Name: `WEBVIEW2_BROWSER_EXECUTABLE_FOLDER`  
Value: `<path to desired Microsoft Edge preview channel>`

> The environment variable will be applied to all apps that use WebView2 on the machine, unlike the registry key, which can be set per app.

Author: ```suggestion
The target app can't be set when using the environment variable approach.  Therefore, when set as a global environment, this setting will affect all apps that use WebView2 on the machine.
```
This comment is a fresh entering of Nic's Outdated [comment](https://github.com/MicrosoftDocs/edge-developer/pull/3053/files/ebb0590c0e011b7e4455bacd633032af0600358e#r1561815133), where Nic asked if it would be more accurate to say it this way.  @champnic @victorhuangwq

Author: Rendered articles for review: 
* **Test upcoming APIs and features** (update to use new API)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel?branch=pr-en-us-3053
* **Prerelease testing** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/prerelease-testing?branch=pr-en-us-3053
* **Self-hosting** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/self-hosting?branch=pr-en-us-3053

For "Outdated" comments, to see latest Markdown/commits, use fresh Files Changed tab:
https://github.com/MicrosoftDocs/edge-developer/pull/3053/files

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_releasechannels)
* [ICoreWebView2EnvironmentOptions7::put_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_releasechannels)
* [COREWEBVIEW2_RELEASE_CHANNELS enum](/microsoft-edge/webview2/reference/win32/webview2-idl#corewebview2_release_channels)

##### [Registry key](#tab/registry-key)

1. Open a PowerShell terminal or PowerShell-enabled command prompt.

1. Modify and then run the following command:

   ```reg
   REG ADD <HKLM/HKCU>\Software\Policies\Microsoft\Edge\WebView2\ReleaseChannels /v WebView2APISample.exe /t REG_SZ /d "0,1"
   ```

Replace `WebView2APISample.exe` with your own app executable name or the application user model ID.  Using a wildcard (*) as the value name will apply the override to _all_ WebView2 apps on the machine, and can result in unexpected behavior.

##### [Environment variable](#tab/environment-variable)

Name: `WEBVIEW2_RELEASE_CHANNELS`  
Value: `<string of desired channels (ex: 0,1)>`

> The environment variable if set globally is applied to all apps that use WebView2 on the machine, unlike the registry key, which can be set per app.

Author: check wording here too; sync w/ elsewhere in this .md file.

Author: Rendered articles for review: 
* **Test upcoming APIs and features** (update to use new API)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel?branch=pr-en-us-3053
* **Prerelease testing** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/prerelease-testing?branch=pr-en-us-3053
* **Self-hosting** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/self-hosting?branch=pr-en-us-3053

For "Outdated" comments, to see latest Markdown/commits, use fresh Files Changed tab:
https://github.com/MicrosoftDocs/edge-developer/pull/3053/files

Reviewer: I wouldn't mind per-app as well.

Author: I might: First rewrite the two spots as "per-app", and then enter two fresh comments to possibly sync the wording in both spots, then Resolve present thread.

Author: entered at both "per-app" spots:
```
<!-- todo: there are two "per-app" spots in this .md file.  Make them same?
The environment variable will be applied to all apps that use WebView2 on the machine, unlike the registry key, which can be set per-app.
The environment variable, if set globally, is applied to all apps that use WebView2 on the machine.  In contrast, the registry key can be set per-app.
-->
```

.NET:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environmentoptions.channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/dotnet/api/microsoft.web.webview2.core.corewebview2channelsearchkind)

WinRT:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environmentoptions#channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2channelsearchkind)

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_channelsearchkind)
* [ICoreWebView2EnvironmentOptions7::put_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_channelsearchkind)
* [COREWEBVIEW2_CHANNEL_SEARCH_KIND enum](/microsoft-edge/webview2/reference/win32//webview2-idl#corewebview2_channel_search_kind)

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Author: Rendered articles for review: 
* **Test upcoming APIs and features** (update to use new API)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel?branch=pr-en-us-3053
* **Prerelease testing** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/prerelease-testing?branch=pr-en-us-3053
* **Self-hosting** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/self-hosting?branch=pr-en-us-3053

For "Outdated" comments, to see latest Markdown/commits, use fresh Files Changed tab:
https://github.com/MicrosoftDocs/edge-developer/pull/3053/files

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Reviewer 2: It looks like the docs should be published with the next Release SDK, which is happening this week or next week.

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Author: https://github.com/MicrosoftDocs/edge-developer/pull/3125

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Reviewer: Yes that's correct. We will publish this right after the release notes next week is published.

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Author: ```
<!-- todo: Update these API Ref links to go to stable interfaces, after they are published; 
after April RelNotes is published in week of Monday, April 22, update these links, and then publish the present article/PR.
PR #3125 (Release Notes for April 2024 Release SDK & Prerelease SDK):
https://github.com/MicrosoftDocs/edge-developer/pull/3125
Live RelNotes to check for presence of April RelNotes:
https://learn.microsoft.com/microsoft-edge/webview2/release-notes/
-->
```

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published -->

Author: this comment thread will probably be Outdated.  Create fresh Comment at "todo", then Resolve present comment.

To deploy the preview channels of Microsoft Edge to multiple machines, do either of the following:

* Option 1: Manually install preview channels on machines.

* Option 2: Programmatically deploy preview channels via API endpoints.

These options are explained below.

<!-- ---------- -->
###### Option 1: Manually install preview channels on machines

Insider channels can be manually installed (either manually or through a deployed custom script) through the following enterprise links:

| Channel | Link |
| --- | --- |
| Canary | [Download](https://go.microsoft.com/fwlink/?linkid=2084649&Channel=Canary&language=en) |
| Dev | [Download](https://go.microsoft.com/fwlink/?linkid=2093291) |
| Beta | [Download](https://go.microsoft.com/fwlink/?linkid=2093376) |

Manual installation only needs to be done once per machine.  Prerelease channels are evergreen, so they will automatically get updated when newer versions are available.

<!-- ---------- -->
###### Option 2: Programmatically deploy preview channels via API endpoints

In your app's code, you can write your own custom logic to deploy the latest version of the preview channel of Microsoft Edge by doing the following:

> 1. Call <!-- todo: all of? any of? one of? --> the following API endpoints.

Author: Call all 3 API endpoints, or just one, or some/selected ones?
@victorhuangwq @champnic

Author: Rendered articles for review: 
* **Test upcoming APIs and features** (update to use new API)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel?branch=pr-en-us-3053
* **Prerelease testing** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/prerelease-testing?branch=pr-en-us-3053
* **Self-hosting** (new article)
   * https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/self-hosting?branch=pr-en-us-3053

For "Outdated" comments, to see latest Markdown/commits, use fresh Files Changed tab:
https://github.com/MicrosoftDocs/edge-developer/pull/3053/files

Reviewer 2: I'm not sure. @vickiez ?

Reviewer: @mikehoffms they should call the endpoint corresponding to the preview channel that they want to deploy

Author: ```
###### Option 2: Programmatically deploy preview channels via API endpoints

In your app's code, write your own custom logic to deploy the latest version of a preview channel of Microsoft Edge as follows:

1. Call the following API endpoint that corresponds to the preview channel that you want to deploy:
```

<!-- ------------------------------ -->
#### Forward-compatibility testing by using the preview channel

We recommend that you test upcoming changes using preview channels to ensure your app's forward compatibility, by following these best practices:

* [Prerelease testing using preview channels](./prerelease-testing.md)
* [Self-host by deploying preview channels](./self-hosting.md)

<!-- ====================================================================== -->
## Downloading the prerelease SDK and a preview channel

To use experimental APIs, download a prerelease version of the WebView2 SDK from [Microsoft.Web.WebView2 package](https://www.nuget.org/packages/Microsoft.Web.WebView2).

Preview channels of Microsoft Edge, which include preview WebView2 runtime, are also called _insider channels_.  After being installed, the preview channels will stay up-to-date and automatically install the latest versions available for that channel.  For example, this means the Canary channel will update itself almost daily.  To download a Microsoft Edge preview channel, see [Microsoft Edge Insider Channels](https://www.microsoft.com/edge/download/insider).

> To deploy an app to multiple machines, see [how to deploy preview channels](#how-to-deploy-preview-channels).

Author: per Victor: it should be:
"To deploy the preview channel to multiple machines, see ..."

> To deploy an app to multiple machines, see [how to deploy preview channels](#how-to-deploy-preview-channels).

Author: ```
To deploy the preview channel to multiple machines, see 
[How to deploy preview channels](#how-to-deploy-preview-channels), below.
```

.NET:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environmentoptions.channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/dotnet/api/microsoft.web.webview2.core.corewebview2channelsearchkind)

WinRT:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environmentoptions#channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2channelsearchkind)

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_channelsearchkind)
* [ICoreWebView2EnvironmentOptions7::put_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_channelsearchkind)
* [COREWEBVIEW2_CHANNEL_SEARCH_KIND enum](/microsoft-edge/webview2/reference/win32/webview2-idl#corewebview2_channel_search_kind)

> <!-- todo: test API Ref links after April RelNotes is published ~Apr 22

Author: todo: test API Ref links after April RelNotes is published ~Apr 22
PR #3125 (Release Notes for April 2024 Release SDK & Prerelease SDK):
https://github.com/MicrosoftDocs/edge-developer/pull/3125
Live RelNotes to check for presence of April RelNotes:
https://learn.microsoft.com/microsoft-edge/webview2/release-notes/

Author: ready to test now.  click links in rendered PR

Author: passed.

* Use the built-in policy on Intune.  See the following articles.  The steps for WebView2 are the same as for Microsoft Edge, except use the category "Microsoft Edge WebView2" instead of "Microsoft Edge".
   * [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge) in the Microsoft Intune documentation.
   * [Configure Microsoft Edge policy settings with Microsoft Intune](/deployedge/configure-edge-with-intune) in the Microsoft Edge Enterprise documentation.

---

<!-- ====================================================================== -->
## Setting the browser executable folder (for local testing)

An alternative approach is to use a browser executable folder.  In this approach, you specify a folder that contains runtime binaries.  This folder can be any of the following locations:
*  The installed location of the WebView2 Runtime.
*  A preview channel of Microsoft Edge.

> *  A folder containing fixed version binaries that you have deployed to the machine yourself.

Reviewer 2: ```suggestion
*  A folder containing Fixed Version binaries that you have deployed to the machine yourself.
```

* Use the built-in policy on Intune.  See [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge).

---

<!-- ====================================================================== -->
## FAQs

<!-- ------------------------------ -->
#### When does the new runtime get picked up?

If you are using an environment variable, registry key, or group policy, they are only applied to any new WebView2 Runtime browser processes that are created; they do not apply to WebView2 Runtime browser processes that are already running.  For example, if you already have a WebView2 with a specific user data folder (and therefore a corresponding browser process) and change `ChannelSearchKind` before attempting to create a new WebView2 using that same user data folder, it will use the existing WebView2 Runtime browser process, and not search for a different runtime.

To ensure that all processes use the environment variable, registry key, or group policy, you need to do any of the following:
* Restart the app.

> * Re-create the WebView2 controls.

Reviewer 2: ```suggestion
* Re-create all of the WebView2 controls sharing a user data folder.
```

Author: ```
* Re-create all of the WebView2 controls that share a user data folder.
```
review this wording, @champnic

* Create a new WebView2 that has a different user data folder than any other currently running WebView2 controls.

<!-- ------------------------------ -->

> #### When using `ChannelSearchKind`, how do I make sure the right channel is picked up when I have multiple preview browser channels?

Reviewer 2: This is a really big section that feels like it doesn't belong in the FAQ. Instead we should probably have a link above in the "Channel Search Kind" that's like "If you need a more specific order, checkout this section + <link>".

Author: trying latest heading:
```
## How to use `ChannelSearchKind` to ensure that the correct preview browser channel is used
```

Author: Link from which section to which section?  Outline of article:
```
# Test upcoming APIs and features
#### Use a preview channel of WebView2 to test your app against the most recent experimental WebView2 APIs
#### Interplay between the WebView2 Prerelease SDK and Microsoft Edge Preview Channels
#### Forward-compatibility testing by using the preview channel

## Downloading the prerelease SDK and a preview channel

## Approaches to making your app use a specific browser channel

## Switching the channel search order (recommended)
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## Setting the browser executable folder (for local testing)
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## When the new runtime gets picked up

## How to use `ChannelSearchKind` to ensure that a particular preview browser channel is used
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## How to set an environment variable
#### Setting environmental variables by using APIs, for a specific app
#### Setting environmental variables globally, for all apps

## Deploy preview channels
#### Option 1: Manually install preview channels on machines
#### Option 2: Programmatically deploy preview channels via API endpoints

## See also
```
@champnic @victorhuangwq

Author: new outline after moved the section up:
```
# Test upcoming APIs and features
#### Use a preview channel of WebView2 to test your app against the most recent experimental WebView2 APIs
#### Interplay between the WebView2 Prerelease SDK and Microsoft Edge Preview Channels
#### Forward-compatibility testing by using the preview channel

## Downloading the prerelease SDK and a preview channel

## Approaches to making your app use a specific browser channel

## Switching the channel search order (recommended)
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## How to use `ChannelSearchKind` to ensure that a particular preview browser channel is used
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## Setting the browser executable folder (for local testing)
##### [API](#tab/api)
##### [Registry key](#tab/registry-key)
##### [Environment variable](#tab/environment-variable)
##### [Group policy](#tab/group-policy)

## When the new runtime gets picked up

## How to set an environment variable
#### Setting environmental variables by using APIs, for a specific app
#### Setting environmental variables globally, for all apps

## Deploy preview channels
#### Option 1: Manually install preview channels on machines
#### Option 2: Programmatically deploy preview channels via API endpoints

## See also
```

Name: `WEBVIEW2_RELEASE_CHANNELS`  <!-- 2 trailing spaces = linebreak -->
Value: `<string of desired channels (ex: 0,1)>`

The target app can't be set when using the environment variable approach.  Therefore, when set as a global environment, this setting will affect all apps that use WebView2 on the machine.

##### [Group policy](#tab/group-policy)

Set the `ReleaseChannel` policy.

* Name: `<app exe name or app user model ID - ex: WebView2APISample.exe>`  <!-- 2 trailing spaces = linebreak -->
* Value: `<string of desired channels (ex: 0,1)>`

Do either of the following:

* Download the Microsoft Edge policy files, which include the WebView2 policy files, from [Download and configure Microsoft Edge for Business](https://www.microsoft.com/edge/business/download).  For more information, see [Configure Microsoft Edge policy settings on Windows devices](/deployedge/configure-microsoft-edge).

* Use the built-in policy on Intune.  For more information, see [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge).

---

<!-- ------------------------------ -->

> #### How to use an environment variable

Reviewer 2: FAQ should be a question
```suggestion
#### How do I set an environment variable?
```

Reviewer 2: Or maybe this doesn't need to be in FAQ

Author: got rid of FAQ format, promoted child headings, moved toward standard wording of headings ie not questions

Author: ```
## How to set an environment variable
```

> #### How to use an environment variable

Reviewer 2: We should link to this section from each of the "Environment" sections above

Author: 3x:
```
See also:
* [How to set an environment variable](#how-to-set-an-environment-variable), below.
```

<!-- ---------- -->
###### Setting environmental variables using API

> You can set environment variables by using `SetEnvironmentVariable`.

Reviewer 2: ```suggestion
You can set environment variables by using `SetEnvironmentVariable`. If you call this function from your host app, it will limit the scope of the environment variable to just your host app and the WebView2 controls it creates, and won't affect other apps on the system.
```

Author: ```
You can set environment variables by using `SetEnvironmentVariable`.  
Calling this function from your host app limits the scope of the environment variable 
to just your host app and the WebView2 controls that it creates, 
and doesn't affect other apps on the system.
```

* .NET: [Environment.SetEnvironmentVariable Method](/dotnet/api/system.environment.setenvironmentvariable)

* WinRT: [EnvironmentManager.SetEnvironmentVariable Method](/windows/windows-app-sdk/api/winrt/microsoft.windows.system.environmentmanager.setenvironmentvariable)

* Win32: [SetEnvironmentVariable function (winbase.h)](/windows/win32/api/winbase/nf-winbase-setenvironmentvariable)

<!-- ---------- -->

> ###### Setting environmental variables globally

Reviewer 2: ```suggestion
###### Setting environmental variables globally

You can also set environment variables by using the Windows system UI. If you do, it will make the scope of the environment variable global for the current user or system, and will affect all apps that create WebView2 controls.
```

Author: ```
###### Setting environmental variables globally

You can set environment variables by using the Windows system UI.  
This approach makes the scope of the environment variable global for the current user or system, 
and affects all apps that create WebView2 controls.

1. In the Windows search bar, enter "environment", and then select **Edit the system environment variables**:
```

   The **System Properties** dialog opens.

1. Select the **Advanced** tab, and then click the **Environment Variables** button:

   ![The Environment Variables button in the System Properties dialog](./set-preview-channel-images/system-properties-env-vars.png)

   The **Environment Variables** dialog opens.

1. In the **User variables** section at the top, click the **New** button.  The **New User Variable** dialog opens.

1. In the **Variable name** text box, enter the environment variable name, either `WEBVIEW2_CHANNEL_SEARCH_KIND` or `WEBVIEW2_BROWSER_EXECUTABLE_FOLDER`.

1. In the **Variable value** text box, do either of the following:

   * If setting the `ChannelSearchKind`, enter `1` to reverse the search order from the default, to use the preview channels first (Canary, Beta, then Dev), followed by the WebView2 Runtime.  Any value other than `1` causes the default search order to be used (WebView2 Runtime, Edge Dev, Edge Beta, and then Edge Canary).

   * If setting the `BrowserExecutableFolder`, enter the path to your preferred browser channel or fixed-version binaries.

1. Click the **OK** buttons to close the dialogs.

<!-- ------------------------------ -->

> #### How to deploy preview channels

Reviewer 2: ```suggestion
#### How do I deploy Microsoft Edge preview channels?
```

Reviewer 2: Or maybe this doesn't need to be in FAQ

Author: FAQs lack structure; prefer structured docs rather than FAQ format.

Author: Dissolved FAQ; promoted and standardized its child headings.

ms.topic: conceptual
ms.service: microsoft-edge
ms.subservice: webview
ms.date: 04/19/2024
---

> # Test upcoming APIs and features

Author: added 2nd, verbose See Also link at top of article:
```
See also:
* [Phases of introducing APIs](../concepts/versioning.md#phases-of-introducing-apis) in _Understand the different WebView2 SDK versions_.
* [Understand the different WebView2 SDK versions](../concepts/versioning.md) - Use a Prerelease version of 
the SDK along with a preview channel of Microsoft Edge; 
or use a Release version of the SDK along with the Runtime.
```

.NET:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environmentoptions.channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/dotnet/api/microsoft.web.webview2.core.corewebview2channelsearchkind)

WinRT:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environmentoptions#channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2channelsearchkind)

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_channelsearchkind)
* [ICoreWebView2EnvironmentOptions7::put_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_channelsearchkind)

> * [COREWEBVIEW2_CHANNEL_SEARCH_KIND enum](/microsoft-edge/webview2/reference/win32//webview2-idl#corewebview2_channel_search_kind)

Author: ```suggestion
* [COREWEBVIEW2_CHANNEL_SEARCH_KIND enum](/microsoft-edge/webview2/reference/win32/webview2-idl#corewebview2_channel_search_kind)
```
@victorhuangwq  - i'll fix - this shows the need to test these api links in some way

.NET:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environmentoptions.channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/dotnet/api/microsoft.web.webview2.core.corewebview2channelsearchkind)

WinRT:
* [CoreWebView2EnvironmentOptions.ChannelSearchKind Property](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environmentoptions#channelsearchkind)
* [CoreWebView2ChannelSearchKind Enum](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2channelsearchkind)

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_channelsearchkind)
* [ICoreWebView2EnvironmentOptions7::put_ChannelSearchKind](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_channelsearchkind)
* [COREWEBVIEW2_CHANNEL_SEARCH_KIND enum](/microsoft-edge/webview2/reference/win32//webview2-idl#corewebview2_channel_search_kind)

> <!-- todo: Update these API Ref links to go to stable interfaces, after they are published; 

Author: todo: update API links after RelNotes PR is merged.

Reviewer: The links matches that of the release notes

Author: Nice to do: I'll test links in PR 3125, then delete the todo comment, and Resolve this thread.

Author: Resolving b/c fresh [Comment](https://github.com/MicrosoftDocs/edge-developer/pull/3053/files#r1572832408)

Win32:
* [ICoreWebView2EnvironmentOptions7::get_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#get_releasechannels)
* [ICoreWebView2EnvironmentOptions7::put_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions7#put_releasechannels)
* [COREWEBVIEW2_RELEASE_CHANNELS enum](/microsoft-edge/webview2/reference/win32/webview2-idl#corewebview2_release_channels)

##### [Registry key](#tab/registry-key)

1. Open a PowerShell terminal or PowerShell-enabled command prompt.

1. Modify and then run the following command:

   ```reg
   REG ADD <HKLM/HKCU>\Software\Policies\Microsoft\Edge\WebView2\ReleaseChannels /v WebView2APISample.exe /t REG_SZ /d "0,1"
   ```

Replace `WebView2APISample.exe` with your own app executable name or the application user model ID.  Using a wildcard (*) as the value name will apply the override to _all_ WebView2 apps on the machine, and can result in unexpected behavior.

##### [Environment variable](#tab/environment-variable)

Name: `WEBVIEW2_RELEASE_CHANNELS`  <!-- 2 trailing spaces = linebreak -->
Value: `<string of desired channels (ex: 0,1)>`

> The environment variable, if set globally, is applied to all apps that use WebView2 on the machine.  In contrast, the registry key can be set per-app.

Author: Reword like elsewhere in this file?  Elsewhere in this file reads:
The target app can't be set when using the environment variable approach.  Therefore, when set as a global environment, this setting will affect all apps that use WebView2 on the machine.
@victorhuangwq

Author: ```
The target app can't be set when using the environment variable approach.
Therefore, when set as a global environment, this setting will affect 
all apps that use WebView2 on the machine.
```

* Value: `1`

Use the value `1` for the reversed-search order, or `0` for the default search order.

Do either of the following:

* Download the Microsoft Edge policy files, which include the WebView2 policy files, from [Download and configure Microsoft Edge for Business](https://www.microsoft.com/edge/business/download).  For more information, see [Configure Microsoft Edge policy settings on Windows devices](/deployedge/configure-microsoft-edge).

* Use the built-in policy on Intune.  See the following articles.  The steps for WebView2 are the same as for Microsoft Edge, except use the category "Microsoft Edge WebView2" instead of "Microsoft Edge".
   * [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge) in the Microsoft Intune documentation.
   * [Configure Microsoft Edge policy settings with Microsoft Intune](/deployedge/configure-edge-with-intune) in the Microsoft Edge Enterprise documentation.

---

<!-- ====================================================================== -->
## How to use `ChannelSearchKind` to ensure that a particular channel is used

> When using `ChannelSearchKind`, here's how you can make sure that a certain preview browser channel is used, when you have multiple preview browser channels installed.

Reviewer: ```suggestion
When using `ChannelSearchKind`, here's how you can make sure that a certain preview browser channel is used when you have multiple preview browser channels installed.
```

The Canary channel is recommended, because it offers the latest experimental APIs.

<!-- ------------------------------ -->
#### Interplay between the WebView2 Prerelease SDK and Microsoft Edge Preview Channels

*  **Prerelease SDK:** This version of the WebView2 SDK provides method signatures for experimental APIs, which allow you to write code using the experimental WebView2 APIs in your app.

*  **Microsoft Edge Preview Channels:** These channels (Edge Beta, Edge Dev, or Edge Canary) contain the necessary Microsoft Edge binaries. They allow your app to run and render using the experimental APIs.

For a deeper dive into how SDK versions interact with WebView2 Runtime and Microsoft Edge preview channels, refer to [Understand the different WebView2 SDK versions](../concepts/versioning.md).

<!-- ------------------------------ -->
#### Forward-compatibility testing by using the preview channel

> We recommend that you test the experimental APIs ahead of time, to ensure your app's forward compatibility, by following these best practices:

Reviewer: gotcha. Let me reword this to:

We recommend that you test upcoming changes using preview channels to ensure your app's forward compatibility, by following these best practices:

To download a Microsoft Edge preview channel, see [Microsoft Edge Insider Channels](https://www.microsoft.com/edge/download/insider).

<!-- ====================================================================== -->
## Approaches to making your app use a specific browser channel

When a WebView2 is initialized, it will attempt to find a valid runtime on the machine to use. This can be the WebView2 Runtime, a preview channel of Microsoft Edge, or a specified location containing Fixed Version binaries.  For more information about supported runtimes, see [Distribute your app and the WebView2 Runtime](../concepts/distribution.md).

There are two ways to make your WebView2 app use a specified preview channel of Microsoft Edge:

* [Switching the channel search order (recommended)](#switching-the-channel-search-order-recommended).

* [Setting the browser executable folder (for local testing)](#setting-the-browser-executable-folder-for-local-testing).

These approaches are described below:

<!-- ====================================================================== -->

> ## Switching the channel search order (recommended)

Reviewer 2: Should we move this above BrowserExecutableFolder if it's the recommended approach?

Reviewer: Yup, this is already above BrowserExecutableFolder

> ## Switching the channel search order (recommended)

Reviewer 2: This is also like the third-time we are describing how to do this. We really should just have 1 place properly describe this, and then link to it from the other locations. I recommend this doc be the source of truth, and the other testing docs link here.

Reviewer: deduped the information

Name: `WEBVIEW2_BROWSER_EXECUTABLE_FOLDER`  
Value: `<path to desired Microsoft Edge preview channel>`

The environment variable will be applied to all apps that use WebView2 on the machine, unlike the registry key, which can be set per app.

##### [Group policy](#tab/group-policy)

Set the `BrowserExecutableFolder` policy.

Name: `<app exe name or app user model ID - ex: WebView2APISample.exe>`
Value: `<path to desired Microsoft Edge preview channel>`

Do either of the following:

* Download the Microsoft Edge policy files, which include the WebView2 policy files, from [Download and configure Microsoft Edge for Business](https://www.microsoft.com/edge/business/download).  See [Configure Microsoft Edge policy settings on Windows devices](/deployedge/configure-microsoft-edge).

* Use the built-in policy on Intune.  See [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge).

---

<!-- ====================================================================== -->
## FAQs

<!-- ------------------------------ -->
#### When does the new runtime get picked up?

> If you are using an environment variable, registry key, or group policy, they are only applied to any new processes that are created; they do not apply to processes that are already running.  To ensure that all processes use the environment variable, registry key, or group policy, you would need to restart the processes.

Reviewer 2: ```suggestion
If you are using an environment variable, registry key, or group policy, they are only applied to any new WebView2 Runtime browser processes that are created; they do not apply to WebView2 Runtime browser processes that are already running. For example, if you already have a WebView2 with a specific user data folder (and therefore a corresponding browser process) and change ChannelSearchKind before attempting to create a new WebView2 using that same user data folder, it will use the existing WebView2 Runtime browser process, and not search for a different runtime.  To ensure that all processes use the environment variable, registry key, or group policy, you would need to restart the app, recreate the WebView2 controls, or create a new WebView2 that has a different user data folder than any other currently running WebView2 controls.
```

<!-- ------------------------------ -->
#### When using `ChannelSearchKind`, how do I make sure the right channel is picked up when I have multiple preview browser channels?

> In addition to reversing the channel search order, you can modify the set of release channels, to provide further control over channel selection.  The loader will search for channels on the device according to the channel search kind, but it will only consider channels that are in the specified set of release channels.

Reviewer 2: ```suggestion
In addition to reversing the channel search order, you can modify the set of release channels to include in the search, to provide further control over channel selection.  The loader will search for channels on the device in the order specified by the channel search kind, but it will only consider channels that are in the specified set of release channels.
```

> This option is useful if you would like to target a particular channel, instead of simply choosing the least stable channel.  By default, release channels includes all channels: Stable (0), Beta (1), Dev (2), and Canary (3).

Reviewer 2: ```suggestion
This option is useful if you would like to target a particular channel, instead of simply choosing the most or least stable channel.  By default, the release channels includes all channels: Stable (0), Beta (1), Dev (2), and Canary (3).
```

Win32:
* [ICoreWebView2ExperimentalEnvironmentOptions::get_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2experimentalenvironmentoptions#get_releasechannels)
* [ICoreWebView2ExperimentalEnvironmentOptions::put_ReleaseChannels](/microsoft-edge/webview2/reference/win32/icorewebview2experimentalenvironmentoptions#put_releasechannels)
* [COREWEBVIEW2_RELEASE_CHANNELS enum](/microsoft-edge/webview2/reference/win32/webview2experimental-idl#corewebview2_release_channels)

##### [Registry key](#tab/registry-key)

1. Open a PowerShell terminal or PowerShell-enabled command prompt.

1. Modify and then run the following command:

   ```reg
   REG ADD <HKLM/HKCU>\Software\Policies\Microsoft\Edge\WebView2\ReleaseChannels /v WebView2APISample.exe /t REG_SZ /d "0,1"
   ```

Replace `WebView2APISample.exe` with your own app executable name or the application user model ID.  Using a wildcard (*) as the value name will apply the override to _all_ WebView2 apps on the machine, and can result in unexpected behavior.

##### [Environment variable](#tab/environment-variable)

Name: `WEBVIEW2_RELEASE_CHANNELS`  
Value: `<string of desired channels (ex: 0,1)>`

> The environment variable is applied to all apps that use WebView2 on the machine, unlike the registry key, which can be set per app.

Reviewer 2: Noted elsewhere - only true if set globally.

##### [Group policy](#tab/group-policy)

Set the `ReleaseChannel` policy.

* Name: `<app exe name or app user model ID - ex: WebView2APISample.exe>`
* Value: `<string of desired channels (ex: 0,1)>`

Do either of the following:

* Download the Microsoft Edge policy files, which include the WebView2 policy files, from [Download and configure Microsoft Edge for Business](https://www.microsoft.com/edge/business/download). Refer to [Configure Microsoft Edge policy settings on Windows devices](/deployedge/configure-microsoft-edge) for more information.

* Use the built-in policy on Intune. Refer to [Configure Microsoft Edge policy settings in Microsoft Intune](/mem/intune/configuration/administrative-templates-configure-edge) for more information.

---

<!-- ------------------------------ -->

> #### How to use an environment variable

Reviewer 2: We should include an example of how to do this programmatically. The below manual instructions are fine to get a single user to change the order for themselves, but most devs would do this programmatically in their app process if this is the approach they decide to take.

Author: Rendered articles for review: 
* **Test upcoming APIs and features** (update to use new API)
   * Internal preview: https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel?branch=pr-en-us-3053
   * [Public preview](https://github.com/MicrosoftDocs/edge-developer/blob/user/mikehoffms/prerelease-testing-self-hosting/microsoft-edge/webview2/how-to/set-preview-channel.md)
   * [Before/Live](https://learn.microsoft.com/microsoft-edge/webview2/how-to/set-preview-channel)
* **Prerelease testing** (new article)
   * Internal preview: https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/prerelease-testing?branch=pr-en-us-3053
   * [Public preview](https://github.com/MicrosoftDocs/edge-developer/blob/user/mikehoffms/prerelease-testing-self-hosting/microsoft-edge/webview2/how-to/prerelease-testing.md)
* **Self-hosting** (new article)
   * Internal preview: https://review.learn.microsoft.com/microsoft-edge/webview2/how-to/self-hosting?branch=pr-en-us-3053
   * [Public preview](https://github.com/MicrosoftDocs/edge-developer/blob/user/mikehoffms/prerelease-testing-self-hosting/microsoft-edge/webview2/how-to/self-hosting.md)

Reviewer: Added

*  A preview channel of Microsoft Edge.
*  A folder containing fixed version binaries that you have deployed to the machine yourself.

If you set the browser executable folder to a specific preview channel of Microsoft Edge, you will need to update the location when that preview channel updates to a newer version. This is because the location includes the version number as part of its path. Therefore, we recommend using this approach for local testing only.

##### [API](#tab/api)

Pass the path to the browser executable folder during environment creation with the following APIs

.NET (WPF):
* [CreationProperties Property](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2.creationproperties)
* [CoreWebView2CreationProperties Class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.corewebview2creationproperties)

.NET (WinForms):
* [CreationProperties Property](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.creationproperties)
* [CoreWebView2CreationProperties Class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.corewebview2creationproperties)

WinRT:

> <!--todo: Unsure if there's a WinRT api, add information here -->

Reviewer: @mikehoffms @vickiez I'm not sure if there's an equivalent for WinRT for the API

Author: @peiche-jessica - is there a WinRT API?

Pass the path to the browser executable folder during environment creation, by using the following APIs.

.NET (WPF):
* [CreationProperties Property](https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.wpf.webview2.creationproperties)
* [CoreWebView2CreationProperties Class](https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.wpf.corewebview2creationproperties)

.NET (WinForms):
* [CreationProperties Property](https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.winforms.webview2.creationproperties)
* [CoreWebView2CreationProperties Class](https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.winforms.corewebview2creationproperties)

WinRT:
* TBD<!--todo: link to WinRT API Ref, if exists -->

Win32:
* [CreateCoreWebView2EnvironmentWithOptions](https://learn.microsoft.com/microsoft-edge/webview2/reference/win32/webview2-idl?#createcorewebview2environmentwithoptions)

Reviewer 3: It's a parameter to [CoreWebView2Environment.CreateWithOptionsAsync](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environment?view=webview2-winrt-1.0.2420.47#createwithoptionsasync)

Author: ```
WinRT:
* [CoreWebView2Environment.CreateWithOptionsAsync Method](/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2environment#createwithoptionsasync)
```
